app_dart/lib/src/service/scheduler.dart

```diff
@@ -888,11 +896,25 @@ class Scheduler {
 
     final checkRunGuard = CheckRun.fromJson(json.decode(conclusion.checkRunGuard!));
 
-    // We're in a pull request and the engine is fully built. We need to reverse look up the PR from the check suite,
-    // which sadly is not available in the check_run data. This could be cached at check_run creation time to avoid
-    // this cost.
-    final int checkSuiteId = checkRunEvent.checkRun!.checkSuite!.id!;
-    final PullRequest? pullRequest = await githubChecksService.findMatchingPullRequest(slug, sha, checkSuiteId);
+    // Look up the PR in our cache first. This reduces github quota and requires less calls.
+    PullRequest? pullRequest;
+    try {
+      pullRequest = await findPullRequestFor(
+        firestoreService,
+        checkRunEvent.checkRun!.id!,
+        checkRunEvent.checkRun!.name!,
+      );
+    } catch (e, s) {
+      log.warning('$logCrumb: unable to find PR in PrCheckRuns', e, s);
+    }
+
+    // We'va failed to find the pull request; try a reverse look it from the check suite.
+    if (pullRequest == null) {
+      final int checkSuiteId = checkRunEvent.checkRun!.checkSuite!.id!;
+      pullRequest = await githubChecksService.findMatchingPullRequest(slug, sha, checkSuiteId);
+    }
+
+    // We cannot make any forward progress. Abandon all hope, Check runs who enter here.
     if (pullRequest == null) {
       throw 'No PR found matching this check_run';
```

Comment:
We could maybe leave the check run ID or some other breadcrumb?

Reply:
Done.